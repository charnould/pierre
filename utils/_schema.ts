import { z } from 'zod'

/**
 * Represents a User schema definition using Zod.
 *
 * This schema validates the structure of a user object with the following properties:
 *
 * - `email`: A string that is trimmed and converted to lowercase.
 * - `role`: An enumerated string that can be one of 'administrator', 'contributor', or 'collaborator'.
 * - `config`: A string that is trimmed.
 * - `password_hash`: A string representing the hashed password of the user.
 */
export const User = z.object({
  email: z.string().trim().toLowerCase(),
  role: z.enum(['administrator', 'contributor', 'collaborator']).catch('collaborator'),
  config: z.string().trim(),
  password_hash: z.string()
})

/**
 * Represents a parsed user schema that extends the base `User` schema
 * by adding a `config` property. The `config` property is an array
 * of strings, allowing for additional configuration options.
 *
 * This schema can be used to validate and type-check user data
 * with additional configuration details.
 */
export const Parsed_User = User.extend({ config: z.array(z.string()) })

//
// Incoming SMS parsing schema
export const SMS = z
  .object({
    role: z.string(),
    config: z.string(),
    conv_id: z.string(),
    phone: z.string().nullable(),
    to: z.string(),
    content: z.string().trim()
  })
  .or(z.null())

//
// `./assets/Config` schema
export const Config = z
  .object({
    id: z.string(),
    display: z.string(),
    show: z.array(z.string()),

    custom_data: z.object({ format: z.function() }).or(z.object({})),

    api: z
      .array(
        z.object({
          key: z.enum(['WEBHOOK_KEY_1', 'WEBHOOK_KEY_2', 'WEBHOOK_KEY_3']),
          url: z.string(), // URL check is made elsewhere
          format: z.function().args(
            z.object({
              custom_data: z.array(z.string()),
              content: z.string(),
              role: z.string()
            })
          )
        })
      )
      .default([]),

    models: z
      .object({
        augment_with: z.string(),
        rerank_with: z.string(),
        answer_with: z.string()
      })
      .default({
        augment_with: "openai('gpt-4o-mini-2024-07-18')",
        rerank_with: "openai('gpt-4o-mini-2024-07-18')",
        answer_with: "openai('gpt-4o-mini-2024-07-18')"
      }),
    phone: z.string().nullable(),
    protected: z.boolean(),
    knowledge: z.object({
      community: z.boolean(),
      proprietary: z.object({
        public: z.boolean(),
        private: z.boolean()
      })
    }),
    audience: z.string(),
    persona: z.string(),
    greeting: z.array(z.string()),
    examples: z.array(z.string()),
    disclaimer: z.string().nullable()
  })
  .strict()

//
// Reflects datastore database schema
export const Reply = z.object({
  // Globals
  conv_id: z.string(),
  config: z.string().or(Config),
  role: z.enum(['assistant', 'user', 'system']).default('user'),
  timestamp: z.string().datetime({ offset: true }).nullish().default(null),
  content: z.string(),
  // Metadata
  metadata: z
    .object({
      // User
      user: z.string().trim().toLowerCase().nullish().default(null),
      // Topics
      topics: z.string().trim().toLowerCase().nullish().default(null),
      // Evaluation + satisfaction
      evaluation: z
        .object({
          // CSAT
          customer: z
            .object({
              score: z.coerce.number().nullish().default(null).catch(null),
              comment: z.string().nullish().default(null)
            })
            .default({}),
          // Social housing organization satisfaction
          organization: z
            .object({
              score: z.coerce.number().nullish().default(null).catch(null),
              comment: z.string().nullish().default(null)
            })
            .default({}),
          // AI generated customer satisfaction (CSAT)
          ai: z
            .object({
              score: z.coerce.number().nullish().default(null).catch(null),
              comment: z.string().nullish().default(null)
            })
            .default({})
        })
        .default({}),
      // LLM usage
      tokens: z
        .object({
          // cache : z.number().nullish().default(null),
          prompt: z.number().nullish().default(null),
          completion: z.number().nullish().default(null),
          total: z.number().nullish().default(null)
        })
        .default({})
    })
    .default({})
})

//
// Structured JSON LLM must output for each request
export const Augmented_Query = z.object({
  lang: z.string(),
  contains_profanity: z.boolean(),
  standalone_questions: z.array(z.string()),
  bm25_keywords: z.array(z.string()),
  search_queries: z.array(z.string())
})

//
// AI Context
export const AIContext = Reply.extend({ query: Augmented_Query.nullable().default(null) })
  .merge(
    z.object({
      chunks: z
        .object({
          community: z.array(z.string()).nullish().default([]),
          private: z.array(z.string()).nullish().default([]),
          public: z.array(z.string()).nullish().default([])
        })
        .default({
          community: [],
          private: [],
          public: []
        }),
      custom_data: z.object({ raw: z.array(z.string()), transformed: z.string().default('') }),
      conversation: z
        .array(
          z.object({
            role: z.enum(['assistant', 'user', 'system']),
            content: z.string()
          })
        )
        .default([])
    })
  )
  .refine(async (c) => {
    // Change config name for config content
    if (typeof c.config === 'string') {
      c.config = (await import(`../assets/${c.config}/config`)).default
    }

    // Format data to be the one wanted by API
    if (typeof c.config !== 'string' && 'format' in c.config.custom_data) {
      if (
        Array.isArray(c.custom_data.raw) &&
        c.custom_data.raw.length === 1 &&
        c.custom_data.raw[0] === ''
      ) {
        c.custom_data.transformed = ''
      } else {
        c.custom_data.transformed = c.config.custom_data.format(c.custom_data.raw) as string
      }
    }

    return true
  })

//
//
export type SMS = z.infer<typeof SMS>
export type User = z.infer<typeof User>
export type Parsed_User = z.infer<typeof Parsed_User>
export type Reply = z.infer<typeof Reply>
export type Config = z.infer<typeof Config>
export type AIContext = z.infer<typeof AIContext>
export type Augmented_Query = z.infer<typeof Augmented_Query>
