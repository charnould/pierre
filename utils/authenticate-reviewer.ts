import type { Context, Next } from 'hono'
import { getSignedCookie } from 'hono/cookie'

export const authenticate = async (c: Context, next: Next) => {
  const cookie = await getSignedCookie(c, Bun.env.AUTH_SECRET as string, 'pierre')
  if (cookie === undefined) c.set('is_auth', false)
  else c.set('is_auth', true)
  await next()
}
