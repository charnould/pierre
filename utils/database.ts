import { Database } from 'bun:sqlite'
import * as sqlite_vec from 'sqlite-vec'
import { z } from 'zod'

// Builtin SQLite library on MacOS doesn't allow extensions
Database.setCustomSQLite('/opt/homebrew/opt/sqlite/lib/libsqlite3.dylib')

// Database Factory Function
//
// This function initializes and returns a database connection based on the
// specified database name. It supports multiple predefined database names, each
// pointing to a specific SQLite file. For vector databases, the `sqlite_vec`
// module is loaded to provide vector search.
export const db = (db_name: Db_Name): Database | (Database & sqlite_vec.Db) => {
  if (db_name === 'community') {
    const db = new Database('./knowledge/data.sqlite')
    sqlite_vec.load(db)
    return db
  }

  if (db_name === 'proprietary.private') {
    const db = new Database(`./datastores/${Bun.env.SERVICE}/proprietary.private.sqlite`)
    sqlite_vec.load(db)
    return db
  }

  if (db_name === 'proprietary.public') {
    const db = new Database(`./datastores/${Bun.env.SERVICE}/proprietary.public.sqlite`)
    sqlite_vec.load(db)
    return db
  }

  if (db_name === 'datastore') {
    const db = new Database(`./datastores/${Bun.env.SERVICE}/datastore.sqlite`)
    return db as Database
  }

  throw new Error('Invalid db_name. Database not initialized.')
}

// Zod schema/TS type
const Db_Name = z.enum(['community', 'proprietary.private', 'proprietary.public', 'datastore'])
export type Db_Name = z.infer<typeof Db_Name>
