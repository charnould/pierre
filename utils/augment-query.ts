import type { CoreMessage } from 'ai'
import dedent from 'dedent'
import { type AIContext, Augmented_Query } from './_schema'
import { convert_to_array } from './convert-to-array'
import { today_is } from './generate-answer'
import { generate_text } from './generate-output'

/**
 * Augments the user's query by analyzing the conversation context and
 * generating various types of information such as language code, location,
 * profanity check, user profile, standalone questions, search queries, BM25
 * keywords, step-back questions, and Hyde answers.
 *
 * @param {AIContext} context - The context of the AI conversation, including
 * the conversation history and configuration.
 * @returns {Promise<Augmented_Query>} - A promise that resolves to an augmented
 * query object containing the analyzed and generated information.
 *
 * The function performs the following steps:
 * 1. Starts performance measurement.
 * 2. Defines multiple prompts for generating augmented queries
 * 3. Uses the specified model to generate responses for each prompt.
 * 4. Parses the generated responses into an `Augmented_Query` object.
 * 5. Ends performance measurement and logs the duration.
 * 6. Logs and returns the augmented query object.
 */
export const augment_query = async (context: AIContext) => {
  // Start performance measurement
  performance.mark('start')

  // Multiple prompts used to generate augmented queries. The goal is to enhance
  // query variety and capture different aspects of the user's intent.

  // Prompt checked on february 2 2025
  const lang_and_profanity: CoreMessage[] = [
    {
      role: 'user',
      content: dedent`/no_think 
        Analyze the given sentence and return two results:  
        1. **Profanity Check** – Determine whether the text contains any actual profanity, swear words, slurs, vulgar expressions, or intentionally offensive/inappropriate language. Do not flag neutral, descriptive, or colloquial phrases unless they are widely recognized as offensive in the given language/culture. Return 'true' only if clear profanity is present; otherwise, return 'false'.  
        2. **Language Detection** – Identify the primary language of the sentence and return its ISO 639-1 two-letter code. If multiple languages are present, return the code of the dominant language.  
                
        **Think carefully and output profanity check first, within <profanity> tags, then language detection, within <lang> tags, followed by your reasoning.**
        
        Sentence: “${context.conversation.slice(-1)[0].content}”`
    }
  ]

  // Prompt checked on february 2 2025
  const standalone_questions: CoreMessage[] = [
    ...context.conversation,
    {
      role: 'assistant',
      content: dedent`/no_think 
        Analyze the conversation history and generate a set of standalone questions in French, each rephrasing **last** user inquiry. Question should be clear, self-contained, and directly reflect the user’s intent without introducing any new details or expanding the scope. If there are multiple questions, separate them with a pipe (|). Do not add any introduction, extra text, or formatting. Keep the questions concise and focused solely on the user’s last inquiries. If a question mentions a specific date or time (e.g., 'qui est d’astreinte aujourd’hui'), use today’s date, which is ${today_is()}.`
    }
  ]

  // Prompt checked on february 2 2025
  const search_queries: CoreMessage[] = [
    ...context.conversation,
    {
      role: 'system',
      content: dedent`
      /no_think
        Analyze the entire conversation thoroughly and generate the following:
        
        1. **Hypothetical Document Embeddings** - Two brief and contextually relevant responses in French capturing all essential information likely to be present in the most authoritative search results addressing the user’s final query. Ensure they are comprehensive yet concise, preserving the original intent and meaning, and include relevant geographic details if a location is mentioned. Provide only the reponses, without any introductory phrases, additional explanations or quotes but separated by a pipe (|).
        
        2. **Stepback Questions** - Two thought-provoking step-back questions in French designed to inspire deeper reflection or a broader perspective on the user’s final intent. Ensure the questions are meaningful, directly tied to the context, and relevant to the discussion. Provide only the questions, without any introductory phrases, additional explanations or quotes but separated by a pipe (|).
        
        3. **Web Search Queries** - Two distinct web search queries in French that directly address the user’s final intent. Each query should cover a different aspect of the inquiry to maximize the variety of relevant search results. Provide only the queries, without any introductory phrases, additional explanations or quotes but separated by a pipe (|).
        
        Output Rules:
        - Output MUST be a **single line**.  
        - Output MUST contain **exactly six segments** (two responses + two questions + two queries).  
        - Each segment MUST be separated by '|'.  
        - Do NOT include titles, explanations, quotes, lists, bullet points, markdown, or any formatting.  
        - Do NOT include introductory text.
        - If a question mentions a specific date or time (e.g., 'qui est d’astreinte aujourd’hui'), use today’s date, which is ${today_is}.
        - If a location is referenced, include the appropriate region, department, city or zipcode.  
        
        Final output format example (illustrative only):
        Lorem ipsum | lorem ipsum | lorem ipsum | lorem ipsum | lorem ipsum | lorem ipsum
        `
    }
  ]

  // Prompt checked on february 2 2025
  const bm25_keywords: CoreMessage[] = [
    ...context.conversation,
    {
      role: 'assistant',
      content: dedent` /no_think
        Analyze the user’s final intent and generate a list of five precise, high-impact keywords in French, separated by a pipe (|), optimized for a BM25-based search engine. Focus on essential terms and critical multi-word phrases while filtering out noise such as verbs, adjectives, adverbs, and generic words. Ensure all keywords are orthographically and grammatically correct, prioritizing exact matches for maximum retrieval relevance. Output only the keywords, formatted as instructed, with no additional text or explanation.`
    }
  ]

  // Get LLM model to generate augmented queries
  const model = context.config.models.augment_with

  // Generate augmented queries using the specified model
  const results = await Promise.all([
    generate_text({ model: model, messages: lang_and_profanity, max_tokens: 100 }),
    generate_text({ model: model, messages: search_queries, max_tokens: 350 }),
    generate_text({ model: model, messages: bm25_keywords, max_tokens: 200 }),
    generate_text({ model: model, messages: standalone_questions, max_tokens: 100 })
  ])

  // Parse/validate the generated responses into an Augmented_Query object. Some
  // transformations are applied to the responses to ensure they are in the
  // correct format

  const profanity = extract_tag_value(results[0], 'profanity', false)
  const language = extract_tag_value(results[0], 'lang', 'fr')

  const augmented_query = Augmented_Query.parse({
    // Convert a string with '|' delimiter to an array
    standalone_questions: convert_to_array(results[3]),
    search_queries: convert_to_array(results[1]),
    bm25_keywords: convert_to_array(results[2]),

    // Convert the profanity check to a boolean
    contains_profanity: typeof profanity === 'string' ? !!profanity.includes('true') : false,

    // prettier-ignore
    // Check if the language code is valid, otherwise default to 'fr'
    lang: [ 'ab','aa','af','ak','sq','am','ar','an','hy','as','av','ae','ay','az',
            'bm','ba','eu','be','bn','bh','bi','bs','br','bg','my','ca','ch','ce',
            'ny','zh','cv','kw','co','cr','hr','cs','da','dv','nl','dz','en','eo',
            'et','ee','fo','fj','fi','fr','ff','gl','ka','de','el','gn','gu','ht',
            'ha','he','hz','hi','ho','hu','ia','id','ga','ig','ik','io','is','it',
            'iu','ja','jv','kl','kn','kr','ks','kk','km','ki','rw','ky','kv','kg',
            'ko','ku','kj','la','lb','lg','li','ln','lo','lt','lu','lv','gv','mk',
            'mg','ms','ml','mt','mi','mr','mh','mn','na','nv','nd','ne','ng','nb',
            'nn','no','ii','nr','oc','oj','cu','om','or','os','pa','pi','fa','pl',
            'ps','pt','qu','rm','rn','ro','ru','sa','sc','sd','se','sm','sg','sr',
            'gd','sn','si','sk','sl','so','st','es','su','sw','ss','sv','ta','te',
            'tg','th','ti','bo','tk','tl','tn','to','tr','ts','tt','tw','ty','ug',
            'uk','ur','uz','ve','vi','vo','wa','cy','wo','fy','xh','yi','yo','za','zu']
            .includes(typeof language === 'string' ? language : '') ? language : 'fr'
  })

  // End performance measurement
  performance.mark('end')

  console.log(augmented_query)
  console.debug(`
    -----------------------------------
    augment_query: ${performance.measure('d', 'start', 'end').duration}ms
    -----------------------------------
    `)

  return augmented_query
}

/**
 * Extracts the value of a specified tag from a given response string.
 *
 * @note This function is tested in tests/unit/utils/extract-tag-value.test.ts
 * @param response - The string containing the HTML/XML response.
 * @param tag - The tag whose value needs to be extracted.
 * @param fallback - The fallback value to return if the tag is not found.
 * @returns The extracted tag value in lowercase and trimmed, or the fallback value if the tag is not found.
 */
export const extract_tag_value = (
  response: string,
  tag: string,
  fallback: number | string | boolean | null
) => {
  const regex = new RegExp(`<${tag}>(.*?)</${tag}>`)
  const match = response.match(regex)
  return match ? match[1].trim().toLowerCase() : fallback
}
