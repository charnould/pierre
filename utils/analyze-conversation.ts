import type { ModelMessage } from 'ai'
import { SQL } from 'bun'
import dedent from 'dedent'
import type { Reply } from './_schema'
import { extract_tag_value } from './augment-query'
import { generate_text } from './generate-output'
import { get_conversation, save_topic, score } from './handle-conversation'

const sql = new SQL(`sqlite:datastores/${Bun.env.SERVICE}/datastore.sqlite`)

/**
 * Scores conversations using an AI model.
 *
 * This function retrieves conversation IDs that have no score from the
 * database, then for each conversation, it retrieves the conversation content,
 * uses a language model to determine the score and reasoning, and saves the
 * score back into the database.
 *
 * The scoring process involves:
 * 1. Retrieving all conversation content.
 * 2. Using a language model to evaluate the conversation and determine a score.
 * 3. Saving the score and reasoning back into the database.
 *
 * The score is determined based on how well the AI understood and addressed the
 * user's intent, with the following scale:
 * - 0: Completely failed
 * - 1: Somewhat relevant but major issues
 * - 2: Mostly good but minor flaws
 * - 3: Perfectly understood and responded
 *
 * If the conversation contains only one message, a specific score and comment
 * are assigned.
 *
 * @returns {Promise<void>} A promise that resolves when the scoring process is complete.
 */
export const score = async (): Promise<void> => {
  // Get the `conv_id` of conversations that have no score
  const sql = new SQL(`sqlite:datastores/${Bun.env.SERVICE}/datastore.sqlite`)
  let conv_ids_missing_score = await sql`
    SELECT DISTINCT
      conv_id
    FROM
      telemetry
    WHERE
      json_extract (metadata, '$.evaluation.ai.score') IS NULL;
  `
  conv_ids_missing_score = conv_ids_missing_score.map((row) => row.conv_id)

  // For each conversation:
  // 1. Retrieve all conversation content.
  // 2. Use an LLM to determine the topic.
  // 3. Save the topic back into the database.
  for await (const conv_id of conv_ids_missing_score) {
    const conversation = (await get_conversation(conv_id)) as Reply[]
    const core_messages = conversation.map(({ role, content }) => ({
      role,
      content
    })) as ModelMessage[]

    let score: number | string | null | boolean
    let comment: number | string | null | boolean

    if (core_messages.length === 1) {
      score = -1
      comment = "PIERRE n'a pas répondu à l'utilisateur."
    } else {
      const messages: ModelMessage[] = [
        ...core_messages,
        {
          role: 'assistant',
          content: dedent`
              Evaluate the preceding AI-user conversation and assign a quality score based on how well the AI understood and addressed the user's intent. Follow these precise steps:
              
              1. Carefully read the entire conversation, noting key requests, questions, and AI responses.
              
              2. Analyze whether the AI:
              - Correctly interpreted the user's primary intent and any implicit needs
              - Provided accurate, relevant, and complete information
              - Structured the response appropriately for the context
              - Avoided unnecessary tangents or omissions
              
              3. Assign ONE score from this scale:
              0: Failed completely (misunderstood the request or provided incorrect/harmful information)
              1: Partially addressed the request (major gaps, inaccuracies, or misalignments with intent)
              2: Generally good response (minor improvements possible in accuracy, completeness, or relevance)
              3: Excellent response (perfectly understood intent and provided optimal assistance)
              
              4. Format your response EXACTLY as follows:
              <score>X</score>
              <reasoning>Your 20-75 word justification in French, explaining specifically why you assigned this score, with concrete examples from the conversation.</reasoning>
              
              Ensure proper XML syntax with matching opening and closing tags. The score must be a single digit (0, 1, 2, or 3) between the score tags.`
        }
      ]

      const config = (await import('../assets/default/config')).default
      const model = config.models
      const answer = await generate_text({
        model: model,
        messages: messages,
        max_tokens: 200
      })

      score = extract_tag_value(answer, 'score', null)
      comment = extract_tag_value(answer, 'reasoning', null)
    }

    await score({
      conv_id: conv_id,
      scorer: 'ai',
      score: score,
      comment: comment
    })
  }

  console.info('✅ AI score assignment completed')
  return
}

/**
 * Assigns a topic to conversations that have no assigned topic using AI.
 *
 * This function performs the following steps:
 * 1. Retrieves the `conv_id` of conversations that have no assigned topic from the database.
 * 2. For each conversation, retrieves all conversation content.
 * 3. Uses a language model to determine the topic of the conversation.
 * 4. Saves the determined topic back into the database.
 *
 * The predefined categories for topics are:
 * - 'rent': The discussion pertains to rent payments.
 * - 'life': The conversation is about daily living in the residence.
 * - 'insurance': The topic concerns a home insurance.
 * - 'moving': The discussion revolves around relocation.
 * - 'technical': The conversation involves maintenance requests.
 * - 'contact': The user is inquiring about ways to contact the lessor.
 * - 'chatbot': The user is asking about the chatbot.
 * - 'multiple': More than one of the above categories is equally relevant.
 * - 'unknown': None of the above categories accurately describe the conversation.
 *
 * @returns {Promise<void>} A promise that resolves when the operation is complete.
 */
export const topicize = async (): Promise<void> => {
  // Get the `conv_id` of conversations that have no assigned topic
  let conv_ids_missing_topic = await sql`
    SELECT DISTINCT
      conv_id
    FROM
      telemetry
    WHERE
      json_extract (metadata, '$.topics') IS NULL;
  `
  conv_ids_missing_topic = conv_ids_missing_topic.map((row) => row.conv_id)

  // For each conversation:
  // 1. Retrieve all conversation content.
  // 2. Use an LLM to determine the topic.
  // 3. Save the topic back into the database.
  for await (const conv_id of conv_ids_missing_topic) {
    const conversation = (await get_conversation(conv_id)) as Reply[]
    const core_messages = conversation.map(({ role, content }) => ({
      role,
      content
    })) as ModelMessage[]

    const messages: ModelMessage[] = [
      ...core_messages,
      {
        role: 'assistant',
        content: dedent`
        
        Analyze the full preceding conversation and classify its primary topic into exactly one of the following predefined categories:
        
        - 'rent': The discussion pertains to rent payments, due dates, amounts, late fees, or related financial matters.
        - 'life': The conversation is about daily living in the residence, neighbors, amenities, social aspects, or general experiences.
        - 'insurance': The topic concerns a home insurance certificate, its submission, validity, or related requirements.
        - 'moving': The discussion revolves around relocation, move-in/move-out dates, lease transfers, or associated logistics.
        - 'technical': The conversation involves maintenance requests, repairs, internet issues, heating problems, or any technical difficulties in the residence.
        - 'contact': The user is inquiring about ways to contact the lessor, landlord, or property management.
        - 'social_housing': The user is asking about obtaining social housing, eligibility, application procedures, waiting lists, or related topics.
        - 'chatbot': The user is asking about the chatbot, how responses are generated, or related AI-specific questions.
        - 'multiple': More than one of the above categories is equally relevant.
        - 'unknown': None of the above categories accurately describe the conversation.
        
        Response Instructions:
        - Return only the exact category name (e.g., rent).
        - Do not include explanations, extra words, formatting, or any additional output.
        - Ensure the classification is 100% consistent and deterministic based on the given definitions.
        
        Your classification:`
      }
    ]

    const topic = (await generate_text({ messages: messages, max_tokens: 50 })).toLowerCase().trim()

    const topics = [
      'rent',
      'life',
      'insurance',
      'moving',
      'technical',
      'contact',
      'social_housing',
      'chatbot',
      'multiple',
      'unknown'
    ]

    if (topics.includes(topic)) await save_topic({ conv_id: conv_id, topic: topic })
  }

  console.info('✅ AI topic assignment completed')
  return
}
