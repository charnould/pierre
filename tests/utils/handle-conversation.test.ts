import { Database } from 'bun:sqlite'
import { afterAll, beforeAll, expect, it, setSystemTime } from 'bun:test'
import { AIContext } from '../../utils/_schema'
import { db } from '../../utils/database'
import {
  delete_conversation,
  get_conversation,
  get_conversations_for_review,
  save_reply,
  score_conversation
} from '../../utils/handle-conversation'

//
//
//
//
//
//
beforeAll(() => {
  setSystemTime(new Date('2012-11-10T09:08:07'))
  const tdb = db('telemetry')
  if (tdb instanceof Database) tdb.query('DELETE FROM telemetry').run()
})

//
//
//
//
//
//
it('should insert 3 replies and retrieve 2 conversations', async () => {
  // id_1
  const r_1_1 = await AIContext.parseAsync({
    conv_id: 'id_1',
    config: 'pierre-ia.org',
    role: 'user',
    content: 'Qui es-tu ?'
  })

  const r_1_2 = await AIContext.parseAsync({
    conv_id: 'id_1',
    config: 'pierre-ia.org',
    role: 'assistant',
    content: 'Je suis Pierre !'
  })

  // id_2
  const r_2_1 = await AIContext.parseAsync({
    conv_id: 'id_2',
    config: 'pierre-ia.org',
    role: 'user',
    content: 'Bonjour'
  })

  save_reply(r_1_1, false)
  save_reply(r_1_2, false)
  save_reply(r_2_1, false)

  // Tests
  expect(get_conversation(r_1_1.conv_id)).toMatchSnapshot()
  expect(get_conversation(r_2_1.conv_id)).toMatchSnapshot()
})

//
//
//
//
//
//
it('should score 2 conversations for customer + organization + external', () => {
  score_conversation(
    {
      conv_id: 'id_1',
      scorer: 'customer',
      score: 0,
      comment: "C'est nul pour le Customer"
    },
    false
  )

  score_conversation(
    {
      conv_id: 'id_1',
      scorer: 'organization',
      score: 1,
      comment: "C'est 1 pour l'Organization"
    },
    false
  )

  score_conversation(
    {
      conv_id: 'id_2',
      scorer: 'external',
      score: 2,
      comment: "C'est 2 pour l'External"
    },
    false
  )

  expect(get_conversation('id_1')).toMatchSnapshot()
  expect(get_conversation('id_2')).toMatchSnapshot()
})

//
//
//
//
//
//
it('should insert 2 replies (from 1 conversation) and delete this conversation', async () => {
  // id_3
  const r_3_1 = await AIContext.parseAsync({
    conv_id: 'id_3',
    config: 'pierre-ia.org',
    role: 'user',
    content: 'Conv 3_1'
  })

  const r_3_2 = await AIContext.parseAsync({
    conv_id: 'id_3',
    config: 'pierre-ia.org',
    role: 'assistant',
    content: 'Conv 3_2'
  })

  save_reply(r_3_1, false)
  save_reply(r_3_2, false)

  // Tests
  expect(get_conversation(r_3_1.conv_id)).toMatchSnapshot()
  expect(delete_conversation(r_3_1.conv_id)).toMatchSnapshot()
})

//
//
//
//
//
//
it('should retrieve all conversations', () => {
  expect(get_conversations_for_review()).toMatchSnapshot()
})

//
//
//
//
//
//
afterAll(() => setSystemTime())
