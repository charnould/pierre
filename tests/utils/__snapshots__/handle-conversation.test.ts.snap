// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`should insert 3 replies and retrieve 2 conversations 1`] = `
[
  {
    "completion_tokens": null,
    "config": "pierre-ia.org",
    "content": "Qui es-tu ?",
    "conv_id": "id_1",
    "cus_comment": null,
    "cus_satisfaction": null,
    "ext_comment": null,
    "ext_satisfaction": null,
    "model": "openai('gpt-4o-mini-2024-07-18')",
    "org_comment": null,
    "org_satisfaction": null,
    "prompt_tokens": null,
    "role": "user",
    "timestamp": "2012-11-10T09:08:07Z",
    "total_tokens": null,
  },
  {
    "completion_tokens": null,
    "config": "pierre-ia.org",
    "content": "Je suis Pierre !",
    "conv_id": "id_1",
    "cus_comment": null,
    "cus_satisfaction": null,
    "ext_comment": null,
    "ext_satisfaction": null,
    "model": "openai('gpt-4o-mini-2024-07-18')",
    "org_comment": null,
    "org_satisfaction": null,
    "prompt_tokens": null,
    "role": "assistant",
    "timestamp": "2012-11-10T09:08:07Z",
    "total_tokens": null,
  },
]
`;

exports[`should insert 3 replies and retrieve 2 conversations 2`] = `
[
  {
    "completion_tokens": null,
    "config": "pierre-ia.org",
    "content": "Bonjour",
    "conv_id": "id_2",
    "cus_comment": null,
    "cus_satisfaction": null,
    "ext_comment": null,
    "ext_satisfaction": null,
    "model": "openai('gpt-4o-mini-2024-07-18')",
    "org_comment": null,
    "org_satisfaction": null,
    "prompt_tokens": null,
    "role": "user",
    "timestamp": "2012-11-10T09:08:07Z",
    "total_tokens": null,
  },
]
`;

exports[`should score 2 conversations for customer + organization + external 1`] = `
[
  {
    "completion_tokens": null,
    "config": "pierre-ia.org",
    "content": "Qui es-tu ?",
    "conv_id": "id_1",
    "cus_comment": "C'est nul pour le Customer",
    "cus_satisfaction": 0,
    "ext_comment": null,
    "ext_satisfaction": null,
    "model": "openai('gpt-4o-mini-2024-07-18')",
    "org_comment": "C'est 1 pour l'Organization",
    "org_satisfaction": 1,
    "prompt_tokens": null,
    "role": "user",
    "timestamp": "2012-11-10T09:08:07Z",
    "total_tokens": null,
  },
  {
    "completion_tokens": null,
    "config": "pierre-ia.org",
    "content": "Je suis Pierre !",
    "conv_id": "id_1",
    "cus_comment": "C'est nul pour le Customer",
    "cus_satisfaction": 0,
    "ext_comment": null,
    "ext_satisfaction": null,
    "model": "openai('gpt-4o-mini-2024-07-18')",
    "org_comment": "C'est 1 pour l'Organization",
    "org_satisfaction": 1,
    "prompt_tokens": null,
    "role": "assistant",
    "timestamp": "2012-11-10T09:08:07Z",
    "total_tokens": null,
  },
]
`;

exports[`should score 2 conversations for customer + organization + external 2`] = `
[
  {
    "completion_tokens": null,
    "config": "pierre-ia.org",
    "content": "Bonjour",
    "conv_id": "id_2",
    "cus_comment": null,
    "cus_satisfaction": null,
    "ext_comment": "C'est 2 pour l'External",
    "ext_satisfaction": 2,
    "model": "openai('gpt-4o-mini-2024-07-18')",
    "org_comment": null,
    "org_satisfaction": null,
    "prompt_tokens": null,
    "role": "user",
    "timestamp": "2012-11-10T09:08:07Z",
    "total_tokens": null,
  },
]
`;

exports[`should insert 2 replies (from 1 conversation) and delete this conversation 1`] = `
[
  {
    "completion_tokens": null,
    "config": "pierre-ia.org",
    "content": "Conv 3_1",
    "conv_id": "id_3",
    "cus_comment": null,
    "cus_satisfaction": null,
    "ext_comment": null,
    "ext_satisfaction": null,
    "model": "openai('gpt-4o-mini-2024-07-18')",
    "org_comment": null,
    "org_satisfaction": null,
    "prompt_tokens": null,
    "role": "user",
    "timestamp": "2012-11-10T09:08:07Z",
    "total_tokens": null,
  },
  {
    "completion_tokens": null,
    "config": "pierre-ia.org",
    "content": "Conv 3_2",
    "conv_id": "id_3",
    "cus_comment": null,
    "cus_satisfaction": null,
    "ext_comment": null,
    "ext_satisfaction": null,
    "model": "openai('gpt-4o-mini-2024-07-18')",
    "org_comment": null,
    "org_satisfaction": null,
    "prompt_tokens": null,
    "role": "assistant",
    "timestamp": "2012-11-10T09:08:07Z",
    "total_tokens": null,
  },
]
`;

exports[`should insert 2 replies (from 1 conversation) and delete this conversation 2`] = `
{
  "changes": 2,
  "lastInsertRowid": 0,
}
`;

exports[`should retrieve all conversations 1`] = `
[
  {
    "completion_tokens": null,
    "config": "pierre-ia.org",
    "content": "Qui es-tu ?",
    "conv_id": "id_1",
    "cus_comment": "C'est nul pour le Customer",
    "cus_satisfaction": 0,
    "ext_comment": null,
    "ext_satisfaction": null,
    "model": "openai('gpt-4o-mini-2024-07-18')",
    "org_comment": "C'est 1 pour l'Organization",
    "org_satisfaction": 1,
    "prompt_tokens": null,
    "role": "user",
    "timestamp": "2012-11-10T09:08:07Z",
    "total_tokens": null,
  },
  {
    "completion_tokens": null,
    "config": "pierre-ia.org",
    "content": "Je suis Pierre !",
    "conv_id": "id_1",
    "cus_comment": "C'est nul pour le Customer",
    "cus_satisfaction": 0,
    "ext_comment": null,
    "ext_satisfaction": null,
    "model": "openai('gpt-4o-mini-2024-07-18')",
    "org_comment": "C'est 1 pour l'Organization",
    "org_satisfaction": 1,
    "prompt_tokens": null,
    "role": "assistant",
    "timestamp": "2012-11-10T09:08:07Z",
    "total_tokens": null,
  },
  {
    "completion_tokens": null,
    "config": "pierre-ia.org",
    "content": "Bonjour",
    "conv_id": "id_2",
    "cus_comment": null,
    "cus_satisfaction": null,
    "ext_comment": "C'est 2 pour l'External",
    "ext_satisfaction": 2,
    "model": "openai('gpt-4o-mini-2024-07-18')",
    "org_comment": null,
    "org_satisfaction": null,
    "prompt_tokens": null,
    "role": "user",
    "timestamp": "2012-11-10T09:08:07Z",
    "total_tokens": null,
  },
]
`;
